parser/tools/generate_ast.cc

```diff
@@ -375,6 +483,60 @@ void emitNodeClassfile(ostream &out, NodeDef &node) {
     out << "    return to_string(buf);\n";
     out << "  }" << '\n';
     out << '\n';
+
+    // toRuby
+    out << "  std::string " << node.name << "::toRuby(const core::GlobalState &gs, int tabs) {" << '\n'
+        << "    fmt::memory_buffer buf;" << '\n';
+    out << "    fmt::format_to(buf, \"s(:" << rubyNodeNameForInternalNodeName(node.name) << "\");" << '\n';
+    // Generate fields
+    for (auto &arg : node.fields) {
+        if (arg.type == Loc) {
+            continue;
+        }
+
+        switch (arg.type) {
+            case Name:
+                out << "    fmt::format_to(buf, \", :\" + JSON::escape(" << arg.name << ".data(gs)->show(gs)));\n";
+                break;
+            case Node:
+                out << "    fmt::format_to(buf, \",\");\n";
+                out << "    if (" << arg.name << ") {\n";
+                out << "     fmt::format_to(buf, \"\\n\");\n";
+                out << "     printTabs(buf, tabs + 1);" << '\n';
+                out << "     printNodeRuby(buf, " << arg.name << ", gs, tabs + 1);\n";
+                out << "    } else {\n";
+                out << "      fmt::format_to(buf, \" nil\");\n";
+                out << "    }\n";
+                break;
+            case NodeVec:
+                out << "    for (auto &&a: " << arg.name << ") {\n";
+                out << "      fmt::format_to(buf, \",\");\n";
+                out << "      if (a) {\n";
+                out << "        fmt::format_to(buf, \"\\n\");\n";
+                out << "        printTabs(buf, tabs + 1);" << '\n';
+                out << "        printNodeRuby(buf, a, gs, tabs + 1);\n";
+                out << "      } else {\n";
+                out << "        fmt::format_to(buf, \" nil\");\n";
+                out << "      }\n";
+                out << "    }\n";
+                break;
+            case String:
+                out << "    fmt::format_to(buf, \", \\\"{}\\\"\", " << arg.name << ");\n";
+                break;
+            case Uint:
+                out << "    fmt::format_to(buf, \", {}\", " << arg.name << ");\n";
+                break;
+            case Loc:
+                // quiet the compiler; we skip Loc fields above
+                abort();
```

Comment:
Why have both a `switch` and an `if`? Could the first part of the switch just be

```cpp
switch (arg.type) {
    case Loc:
        continue;
    case Name:
    // ...
}
```